Introduce TargetDamageWarhead + some TS weapon fixes

reaperrr opened the pull request:
This PR makes more TS weapons use inheritance where sensible/applicable and additionally fixes a few more minor issues I noticed while making the changes.

The commits do contain a few changes on top of merely using inheritance, but unlike in previous PRs, this time I learned my lesson and listed all intentional changes in the respective commit messages.

Closes #12636.

pchote commented on mods/ts/weapons/healweapons.yaml:
```diff
@@ -1,25 +1,25 @@
-	Projectile: Bullet
-		Speed: 1c682
+	Projectile: InstantHit
 	Warhead@1Dam: SpreadDamage
```
I think this would be a good time to implement a `TargetActor` warhead that only deals damage to `Target.Actor`s.  I believe this is how many of the weapons in the original game worked? In any case, it makes a lot more sense here than overriding the spread to not spread.

reaperrr replied:
All our projectiles currently pass only Target.FromPos, so the target passed through the impact is  never an actor (that's what my in-game tests suggest, anyway).

So unless I'm missing something, this would require some code gymnastics that would have to be applied to all projectiles, in addition to the new warhead itself.
Unless we just make the warhead simply perform a small-area scan for actor(s) at the impact pos.

AFAIK the original games simply used Spread=0 which is equal to "only deal 'damage' at impact pos with no spread beyond that".

pchote replied:
I had assumed that the entire point of the `InstantHit` projectile was to target a `Target.Actor`.  Why isn't it doing that?

pchote replied:
In principle you should only need to change `args.PassiveTarget` to `args.GuidedTarget` in the `InstantHit` constructor, then pass that directly to the warhead.

reaperrr replied:
At least on my end the idea was to have a projectile that doesn't have/need any speed, angle or graphics calculations.

But even if `InstantHit` passed the target actor, without additional changes the `TargetActor` warhead would only be compatible with that projectile, which doesn't sound great to me, either.

pchote replied:
The way I would do it is to have the warhead check the `Target` type, and if it's an actor then only apply damage to that actor.  If it's a WPos then search for all actors with an overlapping hit shape, and pick the one that is closest (or all?)

reaperrr replied:
Okay, but ` If it's a WPos then search for all actors with an overlapping hit shape, and pick the one that is closest (or all?)` means that it will need a victim scan radius, and it would save some work on both the warhead and lint rule if #12605 was finished first, which means this PR would likely sit idle for a while if you consider it a blocker.
Personally, I'd prefer to just add a TODO here for now, and then deal with the heal weapons as part of a dedicated `TargetDamageWarhead` PR (as they'd make for a good testcase).

pchote replied:
Ok, fine by me.

pchote replied:
Or even better IMO would be a compromise to implement `TargetDamageWarhead` that only supports `TargetType.Actor`, with a TODO for the target scanning.  I'm assuming that the heal weapons restrict targeting to actors, so don't need the extra logic.

reaperrr replied:
Ok, done (in this PR).

reaperrr committed:
Add TargetDamageWarhead

Only deals damage to the actor that was targeted by the carrying
projectile.

Currently only supported by InstantHit projectile.

reaperrr committed:
Pass InstantHit target directly if inaccuracy is zero

This is required to make the TargetDamage warhead deal damage only to a specific target.

reaperrr committed:
Use inheritance and TargetDamageWH for TS heal weapons

reaperrr committed:
Various misc TS weapon fixes

- Falloff of FireballLauncher and Bomb now more like in TS
- Spread of FireballLauncher now matches TS
- Spread of Bomb is now closer to TS (behaviour not entirely identical, because Spread was cell-based and had a ceiling at ~1.5 cells away from impact cell).
- Bomb now destroys Tiberium like in TS
- Visceroid now uses TargetDamage warhead to damage target directly
- Veins now use TargetDamage warhead to damage target directly

reaperrr commented:
Updated. Fixed `Falloff` and `Spread` on missiles and misc weapons, plus some minor other misc weapon fixes (last two commits).

pchote reviewed (approved):
I'm assuming you've done your research on the balance changes, so I haven't tried to audit those deeper than checking for unintentional node removals/additions.  A couple of minor code nits, then :+1:

reaperrr commented on OpenRA.Mods.Common/Projectiles/InstantHit.cs:
```diff
@@ -71,7 +71,11 @@ public void Tick(World world)
 			if (!doneDamage)
 			{
-				args.Weapon.Impact(Target.FromPos(target), args.SourceActor, args.DamageModifiers);
+				if (info.Inaccuracy.Length > 0)
```
I don't know how to handle inaccuracy in that case?

pchote replied:
I might be missing something obvious, but your `args.PassiveTarget` isn't going to change between the constructor and here, so there shouldn't be any difference between applying the inaccuracy in the constructor vs here.

reaperrr replied:
I meant that `target` is currently a WPos, and blockability and inaccuracy are designed around that. And the way I understand your comment, I'm supposed to change `target` to Target instead of WPos?

pchote replied:
Yeah, exactly.

pchote replied:
See https://github.com/pchote/OpenRA/commit/1cd2eab90db59496317f11cc7ad5d8d71182df65 for what I mean (note: I haven't tested it).

reaperrr commented:
Split everything that isn't affected by the TargetDamageWH to #12640 and applied some of the requested style fixes. I still need to figure out how to implement the last one, though.

pchote commented:
Can you please also swap the workaround SpreadDamage projectile on WormJaw for TargetDamageWarhead?

reaperrr committed:
Return empty Enumerable immediately in InstantHit

...rather than allocating a compiler generated enumerable via 'yield'.

reaperrr committed:
Use TargetDamage for D2k WormJaw

reaperrr commented:
Updated.

- moved inaccuracy check to constructor.
- removed `doneDamage`, that was left-over copy-pasta from `LaserZap`
- changed WormJaw warhead to `TargetDamage`

Also, although I guess it's so insignificant that it wouldn't matter either way, I took the approach of @RoosterDragon's #12599 by returning an empty Enumerable. That commit can easily be removed if it's pointless or even has a negative in this particular case.

abcdefg30 commented:
I can still heal multiple infantry in one go:
![multiheal](https://cloud.githubusercontent.com/assets/7704140/22612909/275d3482-ea74-11e6-943c-703a6af778af.gif)
Could someone else recheck this, please?

abcdefg30 commented:
Haha, never mind. I thought you changed the ra heal weapon, not the ts one.
But anyway, could you include a fix for ra as well, as you're already at it?

abcdefg30 commented:
Actually, while I think about it. Better not. It would just create balance discussions, I fear.

abcdefg30 reviewed (approved):
Works fine in ts and worms still attack.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7b6f35ac9ec0a065bf750d46462136e6e5d75c48...480da128f9c2c64ab876366f36ab9ea936987c7e)

reaperrr commented:
> Actually, while I think about it. Better not. It would just create balance discussions, I fear.

Yeah, that's the reason I didn't include that.
